activesupport/lib/active_support/local_store.rb

```diff
@@ -0,0 +1,49 @@
+# frozen_string_literal: true
+
+module ActiveSupport
+  module LocalStore # :nodoc:
+    module FiberLocal
+      def current
+        Thread.current[:active_support_execution_context] ||= {}
+      end
+    end
+
+    module ThreadLocal
+      def current
+        Thread.current.thread_variable_get(:active_support_execution_context) ||
+          Thread.current.thread_variable_set(:active_support_execution_context, {})
+      end
+    end
+
+    ISOLATION_LEVELS = {
+      fiber: FiberLocal,
+      thread: ThreadLocal,
+    }.freeze
+
+    class << self
+      attr_reader :isolation_level
+
+      def isolation_level=(level)
+        mod = ISOLATION_LEVELS.fetch(level)
+        clear if level != isolation_level && respond_to?(:current, true)
+        singleton_class.define_method(:current, mod.instance_method(:current))
+        singleton_class.send(:private, :current)
+        @isolation_level = level
+      end
+
+      def [](key)
+        current[key]
+      end
+
+      def []=(key, value)
+        current[key] = value
+      end
+
+      def clear
+        current.clear
+      end
+    end
+
+    self.isolation_level = :fiber
```

Comment:
If `:fiber` is meant for a server like falcon, and we ship with puma which would need `:thread`, why is this the default here? I know we're overriding it to `:thread` in the Railtie, but I don't understand why `:fiber` is the default here?

Reply:
It's mostly caution, currently most call sites I want to migrate use `Thread.current[]`, so they're fiber local.

That being said, migrating them to Thread local is very unlikely to break anything, so I'm totally open to directly default to thread local.